### PR TITLE
Implement fallback locales

### PR DESCRIPTION
**Pull request by realityking**
Implementation for the upcoming fallback locale feature.

### CODE REVIEW

File: src/Delivery/DynamicEntry.php [hunk 1]
```diff
@@ -69,9 +69,15 @@ public function getContentType()
...
+    /**
+     * @param  string $name
+     * @param  array  $arguments
+     *
+     * @return mixed
+     */
     public function __call($name, $arguments)
     {
-        if (substr($name, 0, 3) !== 'get') {
+        if (0 !== strpos($name, 'get')) {
```
Comment (mariobodemann) [inline src/Delivery/DynamicEntry.php]: So _name_ can be `getter`, `target` or `together`? I think I liked the former solution better ...

File: src/Delivery/DynamicEntry.php [hunk 2]
```diff
@@ -88,51 +94,73 @@ public function __call($name, $arguments)
             $getId = true;
         }
 
-        if ($fieldConfig !== null && !$fieldConfig->isDisabled()) {
-            if (!isset($this->fields->$fieldName)) {
-                if ($fieldConfig->getType() === 'Array') {
-                    return [];
-                }
+        if ($fieldConfig === null || $fieldConfig->isDisabled()) {
+            trigger_error('Call to undefined method ' . __CLASS__ . '::' . $name . '()', E_USER_ERROR);
```
Comment (mariobodemann) [inline src/Delivery/DynamicEntry.php]: I think this error message is missleading. Shouldn't it be something like `disabled or non existing field found`?

File: src/Delivery/DynamicEntry.php [hunk 3]
```diff
@@ -88,51 +94,73 @@ public function __call($name, $arguments)
...
-            }
+            return null;
+        }
 
-            $result = $value->$locale;
-            if ($getId && !($fieldConfig->getType() === 'Link' || ($fieldConfig->getType() === 'Array' && $fieldConfig->getItemsType() === 'Link'))) {
-                trigger_error('Call to undefined method ' . __CLASS__ . '::' . $name . '()', E_USER_ERROR);
-            }
+        if ($getId && !($fieldConfig->getType() === 'Link' || ($fieldConfig->getType() === 'Array' && $fieldConfig->getItemsType() === 'Link'))) {
+            trigger_error('Call to undefined method ' . __CLASS__ . '::' . $name . '()', E_USER_ERROR);
```
Comment (mariobodemann) [inline src/Delivery/DynamicEntry.php]: Same here. It feels like a copy paste error message.

File: src/Delivery/DynamicEntry.php [hunk 4]
```diff
@@ -88,51 +94,73 @@ public function __call($name, $arguments)
...
+        }
 
-            if ($getId && $fieldConfig->getType() === 'Link') {
-                return $result->getId();
+        $value = $this->fields->$fieldName;
+        if (!$fieldConfig->isLocalized()) {
+            $locale = $this->getSpace()->getDefaultLocale()->getCode();
+        } else {
+            while (!isset($value->$locale)) {
+                $locale = $this->getSpace()->getLocale($locale)->getFallbackCode();
```
Comment (mariobodemann) [inline src/Delivery/DynamicEntry.php]: Is the `getSpace` part cached, or will this actually trigger a request?

File: src/Delivery/DynamicEntry.php [hunk 5]
```diff
@@ -88,51 +94,73 @@ public function __call($name, $arguments)
...
+            trigger_error('Call to undefined method ' . __CLASS__ . '::' . $name . '()', E_USER_ERROR);
+        }
 
-            if ($getId && $fieldConfig->getType() === 'Link') {
-                return $result->getId();
+        $value = $this->fields->$fieldName;
+        if (!$fieldConfig->isLocalized()) {
+            $locale = $this->getSpace()->getDefaultLocale()->getCode();
+        } else {
+            while (!isset($value->$locale)) {
```
Comment (mariobodemann) [inline src/Delivery/DynamicEntry.php]: Do you want to think about an endless loop protection? A counter or something, that you don't spin for ever, if the input data is _invalid_?

File: tests/Integration/FallbackLocaleTest.php [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+        $client->reviveJson('{"sys": {"type": "Space","id": "7dh3w86is8ls"},"name": "Fallback Locales","locales": [{"code": "en-US","default": true,"name": "U.S. English","fallbackCode": null},{"code": "af","default": false,"name": "Afrikaans","fallbackCode": "zu-ZA"},{"code": "zu-ZA","default": false,"name": "Zulu (South Africa)","fallbackCode": "en-US"},{"code": "bs-BA","default": false,"name": "Bosnian (Bosnia and Herzegovina)","fallbackCode": "en-US"},{"code": "ne-NP","default": false,"name": "Nepali (Nepal)","fallbackCode": null}]}');
+
+        // then the Content Type
+        $client->reviveJson('{"sys": {"space": {"sys": {"type": "Link","linkType": "Space","id": "7dh3w86is8ls"}},"id": "sampleType","type": "ContentType","createdAt": "2016-06-21T08:53:55.767Z","updatedAt": "2016-06-21T13:16:34.507Z","revision": 5},"displayField": "title","name": "Sample Type","description": "","fields": [{"id": "title","name": "title","type": "Symbol","localized": true,"required": true,"disabled": false,"omitted": false}]}');
+
+        return $client;
+    }
+
+    /**
+     * No ne-NP
```
Comment (mariobodemann) [inline tests/Integration/FallbackLocaleTest.php]: I'm now not anymore sure, if I like the naming of the locales here. Do you mind writing what no ne-NP means?

File: src/Delivery/DynamicEntry.php (hunk 4 above)
Comment (realityking) [inline src/Delivery/DynamicEntry.php]: It's cached by the client.

File: src/Delivery/DynamicEntry.php (hunk 1 above)
Comment (realityking) [inline src/Delivery/DynamicEntry.php]: It's not perfect, but the old one did that too.

File: src/Delivery/DynamicEntry.php (hunk 2 above)
Comment (realityking) [inline src/Delivery/DynamicEntry.php]: `__call()` is the magic method invoked when the method called could not be found. The error message is emulating the one from the PHP engine when a method is missing, as we swallow that error by declaring `__call()`.

**Pull request merged**